File: apps/ios/retailmanagement/ViewModels/ManagerTimesheetsViewModel.swift
//
//  ManagerTimesheetsViewModel.swift
//  retailmanagement
//

import Foundation
import Observation

@MainActor
@Observable
final class ManagerTimesheetsViewModel {
    var pendingEntries: [TimeEntry] = []
    var summary: TimesheetSummaryResponse?
    var isLoading = false
    var isActionLoading = false
    var summaryLoading = false
    var errorMessage: String?

    var periodStart: Date
    var periodEnd: Date

    init() {
        let cal = Calendar.current
        let now = Date()
        let comps = cal.dateComponents([.year, .month], from: now)
        let start = cal.date(from: comps) ?? now
        let nextMonth = cal.date(byAdding: .month, value: 1, to: start) ?? now
        let end = cal.date(byAdding: .second, value: -1, to: nextMonth) ?? now
        self.periodStart = start
        self.periodEnd = end
    }

    func loadPending(storeId: String) async {
        isLoading = true
        errorMessage = nil
        do {
            let response: PaginatedResponse<TimeEntry> = try await NetworkService.shared.get(
                endpoint: "/api/stores/\(storeId)/timesheets",
                queryItems: [
                    URLQueryItem(name: "status", value: "pending"),
                    URLQueryItem(name: "page_size", value: "100"),
                ]
            )
            pendingEntries = response.data.filter { $0.clockOut != nil }
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func loadSummary(storeId: String) async {
        summaryLoading = true
        errorMessage = nil
        let iso = ISO8601DateFormatter()
        do {
            let response: DataResponse<TimesheetSummaryResponse> = try await NetworkService.shared.get(
                endpoint: "/api/stores/\(storeId)/timesheets/summary",
                queryItems: [
                    URLQueryItem(name: "date_from", value: iso.string(from: periodStart)),
                    URLQueryItem(name: "date_to", value: iso.string(from: periodEnd)),
                ]
            )
            summary = response.data
        } catch {
            errorMessage = error.localizedDescription
        }
        summaryLoading = false
    }

    func updateStatus(storeId: String, entryId: String, status: String) async {
        isActionLoading = true
        do {
            let body = TimeEntryUpdate(status: status)
            let _: DataResponse<TimeEntry> = try await NetworkService.shared.patch(
                endpoint: "/api/stores/\(storeId)/timesheets/\(entryId)", body: body
            )
            await loadPending(storeId: storeId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isActionLoading = false
    }
}

File: apps/android/app/src/main/java/com/retailmanagement/ui/ManagerTimesheetsViewModel.kt
package com.retailmanagement.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.retailmanagement.data.api.RetrofitClient
import com.retailmanagement.data.model.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

class ManagerTimesheetsViewModel : ViewModel() {

    private val api get() = RetrofitClient.api

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isActionLoading = MutableStateFlow(false)
    val isActionLoading: StateFlow<Boolean> = _isActionLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _pendingEntries = MutableStateFlow<List<TimeEntryRead>>(emptyList())
    val pendingEntries: StateFlow<List<TimeEntryRead>> = _pendingEntries.asStateFlow()

    private val _summary = MutableStateFlow<TimesheetSummaryResponse?>(null)
    val summary: StateFlow<TimesheetSummaryResponse?> = _summary.asStateFlow()

    private val _summaryLoading = MutableStateFlow(false)
    val summaryLoading: StateFlow<Boolean> = _summaryLoading.asStateFlow()

    // Default period: current month
    private val cal = Calendar.getInstance()
    val periodStart = MutableStateFlow(startOfMonth())
    val periodEnd = MutableStateFlow(endOfMonth())

    private val isoFmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.getDefault()).also {
        it.timeZone = TimeZone.getTimeZone("UTC")
    }
    private val dateFmt = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

    private fun startOfMonth(): Date {
        cal.set(Calendar.DAY_OF_MONTH, 1)
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    private fun endOfMonth(): Date {
        cal.set(Calendar.DAY_OF_MONTH, cal.getActualMaximum(Calendar.DAY_OF_MONTH))
        cal.set(Calendar.HOUR_OF_DAY, 23); cal.set(Calendar.MINUTE, 59); cal.set(Calendar.SECOND, 59)
        return cal.time
    }

    fun loadPending(storeId: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                val res = api.listTimesheets(storeId, status = "pending", pageSize = 100)
                _pendingEntries.value = res.data.filter { it.clockOut != null }
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isLoading.value = false
        }
    }

    fun loadSummary(storeId: String) {
        viewModelScope.launch {
            _summaryLoading.value = true
            _error.value = null
            try {
                val res = api.getTimesheetSummary(
                    storeId,
                    isoFmt.format(periodStart.value),
                    isoFmt.format(periodEnd.value)
                )
                _summary.value = res.data
            } catch (e: Exception) {
                _error.value = e.message
            }
            _summaryLoading.value = false
        }
    }

    fun updateStatus(storeId: String, entryId: String, status: String, onComplete: () -> Unit) {
        viewModelScope.launch {
            _isActionLoading.value = true
            try {
                api.updateTimesheetEntry(storeId, entryId, TimeEntryUpdate(status))
                loadPending(storeId)
                onComplete()
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isActionLoading.value = false
        }
    }
}
